src/Scale.hs

```diff
@@ -1,99 +1,199 @@
+baseModeIntervals PhrygianDom =
+  nthDegreeIntervals (baseModeIntervals HarmonicMinor) 5
+baseModeIntervals DoubleHarmonicMinor =
+  zipToIntervalSet
+  [Perfect, Major, Minor, Augmented 1, Perfect, Minor, Major] [1..7]
+baseModeIntervals HarmonicMajor =
+  zipToIntervalSet
+  [Perfect, Major, Major, Perfect, Perfect, Minor, Major] [1..7]
+baseModeIntervals DoubleHarmonicMajor =
+  zipToIntervalSet
+  [Perfect, Minor, Major, Perfect, Perfect, Minor, Major] [1..7]
+
+
+modeToIntervals :: Mode -> Set Interval
+modeToIntervals (Mode baseMode exts) = foldr extIntervals (baseModeIntervals baseMode) exts
+  where 
+    extIntervals :: ScaleExt -> Set Interval -> Set Interval
+    extIntervals ext intSet = insert (oldInt <+> (impliedShift $ acc ext)) (delete oldInt intSet)
+      where
+        -- TODO: If there isn't only one interval of a certain degree, the mode is
+        -- ambiguously constructed and we should give a warning.
+        oldInt = elemAt 0 (S.filter (\a -> getSize a == deg ext) intSet)    
+
+
+scaleToNotes :: Scale -> Set Root
+scaleToNotes (Scale root mode) = mapMonotonic (flip jumpIntervalFromNote root) (modeToIntervals mode)
+
+
+modalDistance :: Set Interval -> Set Interval -> Int
+modalDistance mode1 mode2 = sum $ intDistance <$> zip (toAscList mode1) (toAscList mode2)
+  where 
+    intDistance :: (Interval, Interval) -> Int
+    intDistance (i1, i2) = abs $ fromJust $ intervalToDistance (i1 |-| i2) 
+
+
+modesToExts :: Set Interval -> Set Interval -> [ScaleExt]
+modesToExts mode1 mode2 = 
+  let 
+    zippedInts = zip (toAscList mode1) (toAscList mode2)  
+    intervalDiffToAcc :: Interval -> Interval -> Accidental
+    intervalDiffToAcc i1 i2 = shiftToAcc $ fromJust $ intervalToDistance $ i2 |-| i1
+    accToExtList :: Accidental -> Int -> [ScaleExt] -> [ScaleExt]
+    accToExtList accidental degree = 
+      if accidental == natural then
+        id
+      else  
+        (++ [ScaleExt { acc = accidental
+                      , deg = degree 
+                      }])
+  in  
+    foldr (\(i1,i2) exts -> accToExtList (intervalDiffToAcc i2 i1) (getSize i1) exts) 
+          []
+          zippedInts 
+
+
+intervalsToMode :: Set Interval -> [Mode]
+intervalsToMode intSet = 
+  let 
+    sameDegreeModes = filter (\a -> S.map getSize (baseModeIntervals a) 
+                                 == S.map getSize intSet) 
+                      [Lydian ..]
+    distanceFromIntSet :: Set Interval -> BaseMode -> Int
+    distanceFromIntSet iSet mode = modalDistance iSet $ baseModeIntervals mode
+    sortedModes = sortBy (compare `on` distanceFromIntSet intSet) sameDegreeModes
+    exts = modesToExts intSet <$> (baseModeIntervals <$> sortedModes)
+  in
+    takeWhile (\mode -> numAlteredDegsInMode mode == minimum (length <$> exts)) $ (\(x,y) -> Mode x y) <$> zip sortedModes exts
+
+
+isSubsetMode :: Set Interval -> Set Interval -> Bool
+isSubsetMode mode1 mode2 = isSubsetOf mode1 mode2 
```

Comment:
Yeah, I'm not sure this version is particularly useful either. I left it here thinking it might evolve to become the type Mode -> Mode -> Bool, which requires an intermediate conversion to a Set of Intervals and probably warrants a separate function, but I'm not sure yet about how exactly a user might wanna do computation with modes.